pkg/serverless/serverless.go

```diff
@@ -263,58 +291,83 @@ func WaitForNextInvocation(stopCh chan struct{}, daemon *Daemon, metricsChan cha
 		return fmt.Errorf("WaitForNextInvocation: can't unmarshal the payload: %v", err)
 	}
 
-	if payload.EventType == "INVOKE" {
-		log.Debug("Received invocation event")
-		daemon.ComputeGlobalTags(payload.InvokedFunctionArn, config.GetConfiguredTags(true))
-		aws.SetARN(payload.InvokedFunctionArn)
-		daemon.StartInvocation()
-		if coldstart {
-			ready := daemon.WaitUntilClientReady(clientReadyTimeout)
-			if ready {
-				log.Debug("Client library registered with extension")
-			} else {
-				log.Debug("Timed out waiting for client library to register with extension.")
-			}
-			daemon.UpdateStrategy()
-		}
-
-		// immediately check if we should flush data
-		// note that since we're flushing synchronously here, there is a scenario
-		// where this could be blocking the function if the flush is slow (if the
-		// extension is not quickly going back to listen on the "wait next event"
-		// route). That's why we use a context.Context with a timeout `flushTimeout``
-		// to avoid blocking for too long.
-		// This flushTimeout is re-using the forwarder_timeout value.
-		if daemon.flushStrategy.ShouldFlush(flush.Starting, time.Now()) {
-			log.Debugf("The flush strategy %s has decided to flush the data in the moment: %s", daemon.flushStrategy, flush.Starting)
-			flushTimeout := config.Datadog.GetDuration("forwarder_timeout") * time.Second
-			ctx, cancel := context.WithTimeout(context.Background(), flushTimeout)
-			daemon.TriggerFlush(ctx, false)
-			cancel() // free the resource of the context
-		} else {
-			log.Debugf("The flush strategy %s has decided to not flush in the moment: %s", daemon.flushStrategy, flush.Starting)
-		}
-		daemon.WaitForDaemon()
+	if payload.EventType == Invoke {
+		invoke(daemon, payload.InvokedFunctionArn, payload.DeadlineMs, safetyBufferTimout, coldstart, handleInvocation)
 	}
-	if payload.EventType == "SHUTDOWN" {
+	if payload.EventType == Shutdown {
 		log.Debug("Received shutdown event. Reason: " + payload.ShutdownReason)
-
-		if strings.ToLower(payload.ShutdownReason) == "timeout" {
+		isTimeout := strings.ToLower(payload.ShutdownReason.String()) == Timeout.String()
+		if isTimeout {
 			metricTags := addColdStartTag(daemon.extraTags)
 			sendTimeoutEnhancedMetric(metricTags, metricsChan)
 		}
-
-		daemon.Stop()
+		daemon.Stop(isTimeout)
 		stopCh <- struct{}{}
 	}
 
 	return nil
 }
 
+func invoke(daemon *Daemon, arn string, deadlineMs int64, safetyBufferTimout time.Duration, coldstart bool, invocationHandler InvocationHandler) {
+	timeout := computeTimeout(deadlineMs, safetyBufferTimout)
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	doneChanel := make(chan bool)
+	go invocationHandler(doneChanel, daemon, arn, coldstart)
+	select {
+	case <-ctx.Done():
+		log.Debug("Timeout detected, finishing the current invocation now to allow receiving the SHUTDOWN event")
+		daemon.FinishInvocation()
+		return
+	case <-doneChanel:
+		return
+	}
```

Comment:
I like this solution for handling timeouts -- it's easy to understand how this will work.

Reply:
thanks!